Keep draining flag after assignment is revoked in ETS table

slashmili opened the pull request:
In a case that there are lots of partitions and polls are already scheduled, when :drain_after_revoke message is put into the producer's mailbox, it won't be processed until all the poll messages are handle.

With this change, poll messages will be skipped and let the producer to join the rebalancing party as soon as the offset of in-flight message are committed

This PR address part of #92.


I'm not sure about few things:
* The way that I used ETS
* I moved the logic of drain flag to a module.
* Couldn't think of a test scenario yet.

josevalim commented:
Notes:

1. You don't need a separate module, please create functions into the current module

2. Could it happen that drain_after_revoke is called again before this revoke terminates? If so, the ets table should not be a boolean. Instead, you can create a reference with `make_ref` and store it in the ETS table. Then you send the reference as part of GenServer.call and remove it from the table after you drain. You should avoid polling whenever the table is not empty.

slashmili commented:
> 1. You don't need a separate module, please create functions into the current module

Fixed 

> 2. Could it happen that drain_after_revoke is called again before this revoke terminates? 

I believe it's not possible because [brod_group_coordinator:stabilize/3](https://github.com/kafka4beam/brod/blob/master/src/brod_group_coordinator.erl#L492) is only called in the `brod_group_coordinator` GenServer and looking at the code and how we implemented [BroadwayKafka.Producer.assignments_revoked/1](https://github.com/dashbitco/broadway_kafka/blob/main/lib/broadway_kafka/producer.ex#L488) it's a `call` operation and gets blocked until the response is returned back.

slashmili commented:
BTW in a local test scenario with 200 partitions, the assignment_revoke used to take 200k ms and now it takes 795 ms

We are putting this fix into production, and let it run for few days. I'll report back.

josevalim commented on lib/broadway_kafka/producer.ex:
```diff
@@ -680,4 +701,33 @@ defmodule BroadwayKafka.Producer do
+  end
+
+  defp is_draining_after_revoke?(table_name) do
+    [{:draining, value}] = :ets.lookup(table_name, :draining)
+
+    value
```
```suggestion
    :ets.lookup(table_name, :draining, 2)
```

josevalim commented on lib/broadway_kafka/producer.ex:
```diff
@@ -680,4 +701,33 @@ defmodule BroadwayKafka.Producer do
+  end
+
+  defp drain_after_revoke_table_init!(table_name) do
+    table_name =
+      :ets.new(table_name, [
+        :named_table,
+        :public,
+        :set
+      ])
```
```suggestion
    table_name = :ets.new(table_name, [:named_table, :public, :set])
```

josevalim commented on lib/broadway_kafka/producer.ex:
```diff
@@ -304,14 +312,16 @@ defmodule BroadwayKafka.Producer do
 
-    if not state.shutting_down? and state.revoke_caller == nil and offset != nil do
+    if not state.shutting_down? and
+         is_draining_after_revoke?(state.draining_after_revoke_flag) == false and
```
```suggestion
         not is_draining_after_revoke?(state.draining_after_revoke_flag) and
```

josevalim reviewed (commented):
Sounds great then! I have added a few nitpicks and we can ship it!

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
